analytic_engine/src/sst/parquet/hybrid.rs

```diff
@@ -0,0 +1,441 @@
+    array: ArrayRef,
+}
+
+/// Convert timestamp/non key columns to list type
+pub fn build_hybrid_arrow_schema(
+    timestamp_idx: usize,
+    non_key_column_idxes: Vec<usize>,
+    schema: &Schema,
+) -> ArrowSchemaRef {
+    let arrow_schema = schema.to_arrow_schema_ref();
+    let new_fields = arrow_schema
+        .fields()
+        .iter()
+        .enumerate()
+        .map(|(idx, field)| {
+            if idx == timestamp_idx || non_key_column_idxes.contains(&idx) {
+                Field::new(
+                    field.name(),
+                    DataType::List(Box::new(Field::new(
+                        LIST_ITEM_NAME,
+                        field.data_type().clone(),
+                        true,
+                    ))),
+                    true,
+                )
+            } else {
+                field.clone()
+            }
+        })
+        .collect::<Vec<_>>();
+    Arc::new(ArrowSchema::new_with_metadata(
+        new_fields,
+        arrow_schema.metadata().clone(),
+    ))
+}
+
+/// ListArrayBuilder is used for concat slice of different Arrays represented by
+/// ArrayHandle into one ListArray
+struct ListArrayBuilder {
+    datum_kind: DatumKind,
+    list_of_arrays: Vec<ArrayHandle>,
+}
+
+impl ListArrayBuilder {
+    fn new(datum_kind: DatumKind, list_of_arrays: Vec<ArrayHandle>) -> Self {
+        Self {
+            datum_kind,
+            list_of_arrays,
+        }
+    }
+
+    fn build_child_data(&self, offsets: &mut MutableBuffer) -> Result<ArrayData> {
+        let data_type_size = self.datum_kind.size().context(VariableLengthType {
+            type_str: self.datum_kind.to_string(),
+        })?;
+        let values_num = self.list_of_arrays.iter().map(|handle| handle.len()).sum();
+        let mut values = MutableBuffer::new(values_num * data_type_size);
+        let mut null_buffer = MutableBuffer::new_null(values_num);
+        let null_slice = null_buffer.as_slice_mut();
+
+        let mut length_so_far: i32 = 0;
+        offsets.push(length_so_far);
+        for array_handle in &self.list_of_arrays {
+            let shared_buffer = array_handle.data_slice();
+            let null_bitmap = array_handle.null_bitmap();
+
+            for (offset, length) in &array_handle.positions {
+                if let Some(bitmap) = null_bitmap {
+                    for i in 0..*length {
+                        if bitmap.is_set(i + *offset) {
+                            bit_util::set_bit(null_slice, length_so_far as usize + i);
+                        }
+                    }
+                } else {
+                    for i in 0..*length {
+                        bit_util::set_bit(null_slice, length_so_far as usize + i);
+                    }
+                }
+                length_so_far += *length as i32;
+                values.extend_from_slice(
+                    &shared_buffer[offset * data_type_size..(offset + length) * data_type_size],
+                );
+            }
```

Comment:
Can we concat the `null_bitmap` and `shared_buffer` directly to the `bitmap` and `values` rather than set them one by one?